### ISSUE TITLE
WPML Compatibility

### ISSUE THREAD

**Issue by pemaxim** [wp-property/wp-property#154]

**Comment by Mashenka** [comment]
Some notes how WPML works, it creates duplicate of property so you can fill it in other language. At the moment when you switch language it shows you all duplicates from another languages and also dynamic attributes from Developer tab are not translatable.
- [x] Add option to change language like in default posts
  ![image](https://cloud.githubusercontent.com/assets/5111163/11558636/d657517c-99c4-11e5-8a3e-4dc91b19d0f3.png)
- [x] Check also that on frontend properties  showing from correct language.
- [x] attributes and property types labels ( Developer Tab on Settings page ) should be also translatable

**Comment by WWGate** [comment]
Hi @pemaxim 

I'm trying to get posts count to each active language, so I declared WP_Query to get post_count.
It says:
**Fatal error**:  Class 'UsabilityDynamics\WPLT\WP_Query' not found in  **wp-property\vendor\libraries\usabilitydynamics\lib-wp-list-table\lib\classes\class-wp-list-table.php**

how can I fix it?

**Comment by pemaxim** [comment]
@WWGate

Before you proceed with the task, you have to:
- Revert all your changes under vendor directory and never touch anything under that directory directly.
- Be sure, you are familiar with [Composer Manager](https://getcomposer.org/), and you understand what is vendor directory and what is being stored there.
- Be sure, you are familiar with [Grunt](http://gruntjs.com/).
- And, finally, regarding your question. Be sure you know what is [namespaces](http://php.net/manual/en/language.namespaces.rationale.php) in PHP.

**Comment by WWGate** [comment]
@pemaxim 
thank you for your notes. I reverted my last commit and I'll get familiar with composer manager right now.

**Comment by WWGate** [comment]
@pemaxim 
**Check also that on frontend properties showing from correct language.**
currently "Property Overview" shortcode display all items without language filter.
I need to modify the query to add this filter so I checked **get_properties** function in class-functions.php.
I found that it accept additional_sql conditions but I need to make LEFT JOIN with wp_icl_translations,
how can I  do this?

here is the query we should run to filter by language

``` PHP
$query = "SELECT * from wp_posts LEFT JOIN wp_icl_translations ON 
 (wp_posts.ID = wp_icl_translations.element_id)
 WHERE wp_icl_translations.language_code ='". ICL_LANGUAGE_CODE."'";

```

**Comment by pemaxim** [comment]
@WWGate 

Regarding get_properties() function. 
It was written in a bit ugly way a long time ago, but it makes the trick. 
It: 
- [loops every criteria](https://github.com/wp-property/wp-property/blob/latest/lib/class_functions.php#L2944)
- prepares separate SQL query for every criteria and runs it one by one. On every new query we include [$matching_ids](https://github.com/wp-property/wp-property/blob/latest/lib/class_functions.php#L3020). So, every new SQL query search narrows.

So, probably, you just need to apply new filter to function, which will allow to setup $matching_ids before we start running SQL queries based on criterias. And somewhere you might need to add your logic (query or whatever else)... 

I believe, you have to add separate class (class-wpml.php), which will contain all logic related to WPML compatibility. You have not to add WPML compatibility directly. Use hooks ( filters and actions ) to extend the WP-Property native logic. So, it will allow us to determine if WPML plugin is activated at all, and if so, - enable WPML functionality.

**Comment by WWGate** [comment]
@Mashenka 
Done, you can check it now

**Comment by Mashenka** [comment]
Several notes.
- [x]  attributes need to be deleted from WPML string if I will delete them from Developer tab
- [x]  There also meta attributes tab 
- [x] I have translated first 5 attributes in to Russian
  Price,Property type,Bedrooms,Bathrooms,Address
  They are still not translated in Russian property - http://workspace.uds.io/wp-admin/post.php?post=8433&action=edit&lang=ru
  The same with property type

**Comment by WWGate** [comment]
I applied new filters to get translation for attributes and property type in draw_stats function.
it pull the translation now
http://workspace.uds.io/property/325-shenango-blvd-mercer-16121/?lang=ru

**Comment by Mashenka** [comment]
notes:
- [x] Terms tab under Developer tab(This functionality of TERMS add-on)
- [x] We have two default taxonomy in wp-property plugin Features and Community Features. They exist without TERMS plugin, but when it enabled you will see them also in Terms tab
  ![image](https://cloud.githubusercontent.com/assets/5111163/11685727/2fbcfcdc-9e8c-11e5-819e-2918a54f168f.png)
- [ ]  name of the groups of attributes should be also translatable
  ![image](https://cloud.githubusercontent.com/assets/5111163/11664171/4221fbf2-9df1-11e5-8bd7-32c53908909f.png)
- [ ] Data Entry of attributes in Developer tab
  ![image](https://cloud.githubusercontent.com/assets/5111163/11687468/e8d45d5e-9e97-11e5-85b1-8f4665827dd3.png)

**Comment by pemaxim** [comment]
@WWGate 
- [x] Fix: https://github.com/wp-property/wp-property/commit/c479c3d305daf42e9e53c0fd6d3ceb81f1384c1d#commitcomment-14899172
- [x] Fix: https://github.com/wp-property/wp-property/commit/0ceea07310e887577df5bdf25ea90d5e142336cd#commitcomment-14899196
- [x] Improvement: https://github.com/wp-property/wp-property/commit/df233c681f1df0c4390918d773d902e139931242#commitcomment-14899341
- [x] Improvement: https://github.com/wp-property/wp-property/commit/86c97f02c3e10faeb3c2e80a7872369be88f3135#commitcomment-14899359
- [x] Improvements: https://github.com/wp-property/wp-property/commit/b839247d6a9056e6b6abd07f9078cf14563bf766#commitcomment-14899476
- [x] Improvement in filtering_matching_ids() method : http://screencast.com/t/frJUb60P

**Comment by Mashenka** [comment]
- [x] Can you check name of groups item - 
  I have translated them here
  http://workspace.uds.io/wp-admin/admin.php?page=wpml-string-translation%2Fmenu%2Fstring-translation.php&context=property-attributes-custom-attributes&search=property+type&em=1#icl-st-toggle-translations

but do not see translation here
http://workspace.uds.io/property/209-church-street-westmoreland-15658/?lang=ru
- [x] also data entry values not translated on property page(I have translated one, two, three here  - http://workspace.uds.io/wp-admin/admin.php?page=wpml-string-translation%2Fmenu%2Fstring-translation.php&context=property-attributes-values-custom-attributes-value#icl-st-toggle-translations)

![image](https://cloud.githubusercontent.com/assets/5111163/12020886/16ea1072-ad92-11e5-9f7a-56f0e174e425.png)

**Comment by Mashenka** [comment]
- [x] Empty place instead of Main group in default theme
  ![image](https://cloud.githubusercontent.com/assets/5111163/12035725/4136ff84-ae50-11e5-8175-cc206881717d.png)
- [x] In Denali do not show Main group also

**Comment by pemaxim** [comment]
- [x] L10n compatibility: https://github.com/wp-property/wp-property/commit/39d286647b473e398879307fe64c4d048b52ca0c#commitcomment-15203620 Be sure that all context where it's necessary is using L10n.

**Comment by Mashenka** [comment]
All themes 
- [x] When disabled all wpml plugins 
  ![image](https://cloud.githubusercontent.com/assets/5111163/12118096/b2826e96-b3d5-11e5-9c25-9941886d8a29.png)

Madison
- [ ] Property type attribute translated as Тип неджиживости, but it shows Объект Type
  ![image](https://cloud.githubusercontent.com/assets/5111163/12118586/5de97a84-b3d8-11e5-85d3-f00a09849bda.png)
  ![image](https://cloud.githubusercontent.com/assets/5111163/12118609/7af4faea-b3d8-11e5-88f0-ff083650e7db.png)

This fixed Max
- [x] Name of groups not translated if I will use our madison theme
  ![image](https://cloud.githubusercontent.com/assets/5111163/12035808/1a5554fa-ae51-11e5-8a04-f4e4645b54e7.png)
- [x] Attributes not translated in Madison theme
- [x] Property types not translated in Madison theme

Default theme
-  [ ] Now I do not see names of Taxanomy in Default theme
  ![image](https://cloud.githubusercontent.com/assets/5111163/12035604/6021b890-ae4f-11e5-99de-82fcc13a606e.png)
- [x] Property types not translated in Default theme

**Comment by srdjan-jcc** [comment]
Property form translation
![spanish property form es wp property 2016-01-27 12-42-47](https://cloud.githubusercontent.com/assets/3109112/12653703/0071ac98-c5f1-11e5-9bd6-309c6eba9f96.png)

How to translate form labels?
(including custom fields, dropdown options)

http://pluginwpproperty.compatibility.wpml.org/property-form/?lang=es
http://pluginwpproperty.compatibility.wpml.org/wp-admin/edit.php?post_type=property&page=page_feps

**Comment by Mashenka** [comment]
@srdjan-jcc 
Fadi worked only with core on first step. But your comment is pertinently

**Comment by Mashenka** [comment]
Max comments:
- [x] property Search
  ![image](https://cloud.githubusercontent.com/assets/5111163/12754713/7063dc8e-c9dd-11e5-8e22-2953a56903ae.png)
- [x] Madison theme - this is from Madiosn localization
  ![image](https://cloud.githubusercontent.com/assets/5111163/12754727/7d1ce22c-c9dd-11e5-882c-bc4bab654030.png)
- [x] Madison theme
  ![image](https://cloud.githubusercontent.com/assets/5111163/12754735/83834aca-c9dd-11e5-93b0-72da80151e5a.png)

**Comment by Mashenka** [comment]
* [ ] Widget Latest Properties show properties from both language
* [ ] Attributes in Widgets Latest Properties and Featured Properties not translated
![image](https://cloud.githubusercontent.com/assets/5111163/12755707/96315500-c9e1-11e5-84a4-3acf9070d1ae.png)

**Comment by Mashenka** [comment]
- [x] Widget Latest Properties show properties from both language
- [x] Attributes in Widgets Latest Properties, Other Properties and Featured Properties not translated
  ![image](https://cloud.githubusercontent.com/assets/5111163/12755707/96315500-c9e1-11e5-84a4-3acf9070d1ae.png)
- [x] Name of widgets are not translatable
  ![image](https://cloud.githubusercontent.com/assets/5111163/12755955/91ab4670-c9e2-11e5-8643-3ce997e4e128.png)

![image](https://cloud.githubusercontent.com/assets/5111163/12755932/79e5e46e-c9e2-11e5-878a-5830bd6e59e1.png)
- [x] when enabled WPML plugin Child properties widget not showing
  ![image](https://cloud.githubusercontent.com/assets/5111163/12756848/9fe382b2-c9e6-11e5-9e37-0c703e573157.png)
- [x] attributes are not translatable in single property map pop up
  ![image](https://cloud.githubusercontent.com/assets/5111163/12757165/dde90e82-c9e7-11e5-82a5-3ff81120fb89.png)
  - [x] Child Property of particular property. The word of is not translatable
    ![image](https://cloud.githubusercontent.com/assets/5111163/12757340/abd88976-c9e8-11e5-916d-f91f4def2789.png)
- [x] Attributes which using for sorting
  ![image](https://cloud.githubusercontent.com/assets/5111163/12757401/f2d239e4-c9e8-11e5-9ed8-709e22c63684.png)
- [x] on Child properties with WPML enabled showing Array for mulicheckbox attribute
  ![image](https://cloud.githubusercontent.com/assets/5111163/12757558/99b2ca4e-c9e9-11e5-87d1-009f1b19eb03.png)
- [x] Attributes are not translated in Property overview in Denali theme(other themes just do not have labels of attributes in overview)
  ![image](https://cloud.githubusercontent.com/assets/5111163/12757741/55667b64-c9ea-11e5-8266-ac5f00ddcf97.png)

**Comment by Mashenka** [comment]
- [x] in overview name of property type not translated
  ![image](https://cloud.githubusercontent.com/assets/5111163/12886419/05e9f744-ce7e-11e5-95f8-59b870960c52.png)

**Comment by Mashenka** [comment]
- [x] when I search on ru page it redirects me on results with en  properties
